// biz9-data-logic/index.js
/*
Copyright 2016 Certified CoderZ
Author: Brandon Poole Sr. (biz9framework@gmail.com)
License GNU General Public License v3.0
Description: BiZ9 Framework: Data-Logic
*/
const {Str,Num,Obj,Log} = require("/home/think1/www/doqbox/biz9-framework/biz9-utility/source");
class Data_Value_Type {
    static COUNT = 'count';
    static ITEMS = 'items';
    static ONE = 'one';
}
class Data_Field {
    static DATE_CREATE='date_create';
    static DATE_SAVE='date_save';
    static GROUP='group';
    static TABLE='table';
    static ID='id';
    static ITEM='item';
    static ITEMS='items';
    static OBJ='obj';
    static PARENT_ID = 'parent_id';
    static PARENT_TABLE='parent_table';
    static USER='user';
    static USER_ID = 'user_id';
    static USER_TABLE='user_table';
    static SOURCE='source';
    static SOURCE_ID='source_id';
    static SOURCE_TABLE='source_table';
    static TITLE='title';
    static TITLE_URL='title_url';
}
class Data_Type {
    //result
    static RESULT_OK = 'resultOK';
    static RESULT_OK_DELETE = 'delete_resultOK';
    static RESULT_OK_DELETE_CACHE = 'delete_cache_resultOK';
    static RESULT_OK_DELETE_COUNT = 'delete_count';
    static RESULT_OK_GROUP_DELETE = 'group_delete_resultOK';
    static RESULT_OK_DELETE_DATABASE = 'delete_database_resultOK';
    //source
    static SOURCE='source';
    static SOURCE_KEY='source_key';
    static SOURCE_TABLE='source_table';
    static SOURCE_ID='source_id';
    static SOURCE_PARENT_ID='source_parent_id';
    //sort_by
    static SORT_BY_ASC='asc';
    static SORT_BY_DESC='desc';
    //data_source
    static DATA_SOURCE_CLIENT="client";
    static DATA_SOURCE_CLIENT_CACHE="client_cache";
    static DATA_SOURCE_DATABASE="database";
    static DATA_SOURCE_NOT_FOUND="not_found";
    static DATA_SOURCE_SERVER="server";
    static DATA_SOURCE_SERVER_CACHE="server_cache";
}
class Data_Title {
    //general
    static FOREIGN='Foreign';
    static GROUP='Group';
    static JOIN='Join';
    static N_A='N/A';
    //source
    static SOURCE_DATABASE='Database';
    static SOURCE_CACHE='Cache';
    static SOURCE_NOT_FOUND='Not-Found';
}
class Data_Table {
    static BLANK='blank_biz';
    static GROUP='group_biz';
    static USER='user_biz';
}
class Data_Logic {
    static get = (table,id,option) => {
        option = !Obj.check_is_empty(option) ? option : {};
        /* options
            - count
            - parent
            - user
            - generate_id
            - data
        */
        let data = {table:table,id:id};
        if(option.data){
            data = Obj.merge(data,option.data);
        }
        if(option.title){
            data[Data_Field.TITLE] = option.title;
            data[Data_Field.TITLE_URL] = Str.get_title_url(option.title);
        }
        if(option.parent){
            data[Data_Field.PARENT_TABLE] = option.parent[Data_Field.TABLE];
            data[Data_Field.PARENT_ID] = option.parent[Data_Field.ID];
        }
        if(option.user){
            data[Data_Field.USER_TABLE] = option.user[Data_Field.TABLE];
            data[Data_Field.USER_ID] = option.user[Data_Field.ID];
        }
        if(option.generate_id){
            data[Data_Field.ID] = Num.get_id(999);
        }
        if(option.count){
            let data_items = [];
            if(!data.title){
                data.title = 'Item';
                data.title_url = Str.get_title_url(data.title);
            }
            for(let a = 1; a < option.count+1; a++){
                const copy = { ...data };
                copy.title = data.title + " " + a;
                copy.title_url = Str.get_title_url(copy.title);
                if(option.generate_id){
                    copy[Data_Field.ID] = Num.get_id(999);
                }
                data_items.push(copy);
            }
            data = data_items;
        }
        return data;
    };
    static get_search = (table,filter,sort_by,page_current,page_size,option) => {
        option = !Obj.check_is_empty(option) ? option : {};
        return {table:table,filter:filter,sort_by:sort_by,page_current:page_current,page_size:page_size,option:option};
    };
    static get_group = (option) => {
        option = !Obj.check_is_empty(option)  ? option : {};
        let value_type  = option.value_type ? option.value_type : Data_Value_Type.ITEMS;
        let field = option.field ? option.field : {};
        let title = option.title ? Str.get_title_url(option.title) : {};
        let page_current = option.page_current ? option.page_current : 1;
        let page_size = option.page_size ? option.page_size : 0;
        let foreigns = option.foreigns ? option.foreigns : [];
        return {value_type:value_type,field:field,title:title,page_current:page_current,page_size:page_size,foreigns:foreigns};
    };
    static get_foreign = (value_type,foreign_table,foreign_field,parent_field,option) => {
        option = !Obj.check_is_empty(option)  ? option : {};
        value_type = value_type ? value_type : Data_Value_Type.ITEMS;
        foreign_table = foreign_table ? foreign_table : {};
        foreign_field = foreign_field ? foreign_field : Data_Type.FIELD_PARENT_ID;
        parent_field = parent_field ? parent_field : parent_field;
        let field = option.field ? option.field : null;
        let title = option.title ? Str.get_title_url(option.title) : Str.get_lower_case(Str.get_plural(Data_Title.FOREIGN));
        let page_current = option.page_current ? option.page_current : 1;
        let page_size = option.page_size ? option.page_size : 0;
        let foreigns = option.foreigns ? option.foreigns : [];
        return {value_type:value_type,foreign_table:foreign_table,foreign_field:foreign_field,parent_field:parent_field,field:field,title:title,page_current:page_current,page_size:page_size,foreigns:foreigns};
    };
    static get_join = (value_type,search,option) => {
        option = !Obj.check_is_empty(option)  ? option : {};
        value_type = value_type ? value_type : Data_Value_Type.ITEMS;
        search = search ? search : Data_Logic.get_search(Data_Type.DATA_BLANK,{},{},1,0);
        let field = option.field ? option.field : {};
        let distinct = option.distinct ? option.distinct : null; //distinct:{field:'title',sort_by:Data_Type.SEARCH_SORT_BY_DESC}
        let title = option.title ? Str.get_title_url(option.title) : Str.get_lower_case(Str.get_plural(Data_Title.JOIN));
        let foreigns = option.foreigns ? option.foreigns : [];
        let page_current = option.page_current ? option.page_current : 1;
        let page_size = option.page_size ? option.page_size : 0;
        return {value_type:value_type,search:search,field:field,title:title,distinct:distinct,foreigns:foreigns,page_current:page_current,page_size:page_size};
    }
    static copy = (table,item,option)=>{
        let copy_item = Data_Logic.get(table,0);
        option = !Obj.check_is_empty(option)  ? option : {};
        const keys = Object.keys(item);
        keys.forEach(key => {
            if(
                key!=Data_Field.ID&&
                key!=Data_Field.TABLE&&
                key!=Data_Field.SOURCE&&
                key!=Data_Field.DATE_CREATE&&
                key!=Data_Field.DATE_SAVE&&
                key!=Data_Field.OBJ&&
                key!=Data_Field.USER&&
                key!=Data_Field.GROUP&&
                key!=Data_Field.ITEM&&
                !Obj.check_is_array(item[key])&&
                Obj.check_is_value(item[key])
            )
            {
                copy_item[key]=item[key];
            }
        });
        return copy_item;
    };
    static get_not_found = (table,id,option) =>{
        option=!Obj.check_is_empty(option)?option:{};
        if(!table){
            table = Data_Table.BLANK;
        }
        if(!id){
            id = 0;
        }
        let item = Data_Logic.get(table,id);
        item.id = 0;
        item.id_key = id;
        item.title = "Item Not Found";
        item.title_url = Str.get_title_url(item.title);
        item.source = Data_Title.SOURCE_NOT_FOUND;
        return item;
    };
}
module.exports = {
    Data_Field,
    Data_Logic,
    Data_Table,
    Data_Title,
    Data_Type,
    Data_Value_Type
};
